test/unit/sqs/index.test.ts

```diff
@@ -22,140 +23,296 @@ describe("SqsQueue tests", () => {
-			done();
+			mockRequestHandler.handle.mockImplementation((context: Context<TestMessage>) => {
+				expect(context.payload).toStrictEqual(testPayload);
+				done();
+			});
+			queue.sendMessage(testPayload);
 		});
-		queue.sendMessage(testPayload);
-	});
 
 
-	test("Queue is restartable", async (done: DoneCallback) => {
+		test("Queue is restartable", async (done: DoneCallback) => {
 
-		const testPayload = generatePayload();
+			const testPayload = generatePayload();
 
-		mockRequestHandler.handle.mockImplementation((context: Context<TestMessage>) => {
-			expect(context.payload).toStrictEqual(testPayload);
-			done();
+			mockRequestHandler.handle.mockImplementation((context: Context<TestMessage>) => {
+				expect(context.payload).toStrictEqual(testPayload);
+				done();
+			});
+
+			queue.stop();
+
+			//delaying to make sure all asynchronous invocations inside the queue will be finished and it will stop
+			await queue.waitUntilListenerStopped();
+
+			queue.start();
+
+			queue.sendMessage(testPayload);
 		});
 
-		queue.stop();
+		test("Message received with delay", (done: DoneCallback) => {
 
-		//delaying to make sure all asynchronous invocations inside the queue will be finished and it will stop
-		await queue.waitUntilListenerStopped();
+			const testPayload = generatePayload();
+			const receivedTime = {time: new Date().getTime()};
 
-		queue.start();
+			mockRequestHandler.handle.mockImplementation((context: Context<TestMessage>) => {
 
-		queue.sendMessage(testPayload);
-	});
+				if(context.payload.msg !== testPayload.msg) {
+					//discard message if it is not our test message
+					return;
+				}
+
+				try {
+					context.log.info("hi");
+					const currentTime = new Date().getTime();
+					expect(currentTime - receivedTime.time).toBeGreaterThanOrEqual(1000);
+					done();
+				} catch (err) {
+					done(err);
+				}
+			});
+			queue.sendMessage(testPayload, 1);
+		});
+
+		test("Message gets executed exactly once", (done: DoneCallback) => {
+
+			const testPayload = generatePayload();
+			const testData: { messageId: undefined | string } = {messageId: undefined};
+
+			mockRequestHandler.handle.mockImplementation((context: Context<TestMessage>) => {
+
+				if(context.payload.msg !== testPayload.msg) {
+					//discard message if it is not our test message
+					return;
+				}
+
+				try {
+					expect(context.payload).toStrictEqual(testPayload);
+
+					if (!testData.messageId) {
+						testData.messageId = context.message.MessageId;
+					} else if (testData.messageId === context.message.MessageId) {
+						done.fail("Message was received more than once");
+					} else {
+						done.fail("Different message on the tests queue");
+					}
+
+				} catch (err) {
+					done(err);
+				}
+			});
+			queue.sendMessage(testPayload);
+
+			//code before the pause
+			setTimeout(function () {
+				if (testData.messageId) {
+					done();
+				} else {
+					done.fail("No message was received");
+				}
+			}, 3000);
```

Comment:
Use [jest timer mocks](https://jestjs.io/docs/timer-mocks) to control time instead of adding timeouts and slowing down tests.

Reply:
We have an external dependency here - localstack sqs queue. Some things we test with delays beaucse it is faster and easier. Not the best practice, but a good tradeoff for the development speed. 

Not going to change it now (unless it become a problem)